docs/madr/decisions/005-policy-matching.md

```diff
@@ -0,0 +1,647 @@
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/3330 
+
+## Context and Problem Statement
+
+Source and destination selectors (S&D) are trying to solve the problem of mapping between policies and connections. 
+But in the real world of Envoy proxies, 
+there is no such thing as "connection". 
+There are "inbounds" and "outbounds". 
+Depending on policy type, S&D could select either outbounds or inbounds; 
+this is hidden from the user and generates ambiguity and complexity:
+* If S&D selects outbounds under the hood, 
+then the "destination" selector is limited by only the "kuma.io/service" tag.
+* If S&D selects inbounds under the hood, 
+then "source" selectors might have some unexpected behavior (TrafficPermissions shadow each other).
+* Kuma selects "the most specific" policy based on S&D ranks, 
+but sometimes it's hard to say which policy is more specific.
+* "source" selectors could be tricky to implement, especially regarding TCP traffic.
+
+The idea of the current proposal is to make the model more transparent. 
+If we want to map Policies to inbounds/outbounds, 
+it should be clear to users 
+what inbound/outbounds Kuma is choosing for the current policy. 
+To make it work, 
+we have to get rid of the existing "source" and "destination" selectors 
+and replace them with "inbound" and "outbound" selectors.
+
+## Considered Options
+
+* targetRef selector 
+* S&D with constraints [1] 
+
+## Decision Outcome
+
+Chosen option: targetRef selector
+
+### Overview
+
+"TargetRef" selector is inspired by [Gateway API Policy Attachment](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/) specification. 
+Even though, Kuma policy matching cannot be compatible with Gateway API Policy Attachment spec at that moment,
+we don't rule out the possibility in the future. 
+
+Each Kuma policy has a single "TargetRef": 
+
+```protobuf
+message TargetRef {
+  // Kind of the referenced resource
+  // +kubebuilder:validation:Enum=Mesh;MeshSubset;Service;ServiceSubset;Proxy;MeshGatewayRoute;HTTPRoute
+  string kind = 1;
+
+  // Name of the referenced resource
+  string name = 2;
+
+  // Tags are used with MeshSubset and ServiceSubset to define a subset of proxies 
+  map<string, string> tags = 3;
+}
+```
+
+for example:
+
+```yaml
+type: UpstreamTimeout
+mesh: mesh-1
+name: timeouts
+spec:
+  targetRef:
+    kind: Service
+    name: backend
+  conf:
+    # timeouts configuration
+```
+
+"TargetRef" can attach policy to different objects:
+* Mesh
+* MeshSubset
+* Service
+* ServiceSubset
+* Proxy
+* MeshGatewayRoute 
+* HTTPRoute
+
+Traffic direction is expressed inside the "conf" section.
+
+If policy configures outbounds, then "conf" section contains "to" array:
+```yaml
+type: UpstreamTimeout
+mesh: mesh-1
+name: timeouts
+spec:
+  targetRef:
+    kind: Service
+    name: backend
+  conf:
+    to:
+      - targetRef:
+          kind: Service
+          name: payments
+        connectTimeout: 10s
```

Comment:
while the previous nesting wasn't required. Adding one here might help so that the backbone of the policies is always the same:

```
spec:
  targetRef: {}
  config: {} # Only actual type that differs.
  from:
      - targetRef: {}
        config: {} # Only actual type that differs.
  to:
      - targetRef: {}
        config: {} # Only actual type that differs.
```

Reply:
so `spec.config` will be used only for policies that have neither `from` nor `to`, like ProxyTemplate?

Reply:
I added `conf` instead of `config` inside each array's item. Top-level `conf` will be used for example for ProxyTemplates